### PR TITLE
Scalars support

### PR DESCRIPTION
**Pull request by YoelShoshan**
mainly related to adding supports for scalars input/output to our mammal architecture

### CODE REVIEW

File: fusedrug/data/tokenizer/injectortokenizer/injector_tokenizer.py [hunk 1]
```diff
@@ -0,0 +1,207 @@
+from fusedrug.data.tokenizer.modulartokenizer.modular_tokenizer import ModularTokenizer
+from typing import Optional, List, Tuple, Dict
+from tokenizers import Encoding
+import torch
+import re
+from fuse.utils import NDict
+
+
+class InjectorTokenizer(ModularTokenizer):
```
Comment (YoelShoshan) [inline fusedrug/data/tokenizer/injectortokenizer/injector_tokenizer.py:9]: originally I wanted to use this as a drop-in replacement for ModularTokenizer, but in order to avoid code duplication I ended up only storing static methods here.

File: fusedrug/data/tokenizer/ops/modular_tokenizer_ops.py [hunk 2]
```diff
@@ -241,22 +243,30 @@ def __call__(
                 )
 
         if isinstance(data, str):
-            encoded, overflow_info = self._tokenizer.encode(
+            _ans = self._tokenizer.encode(
```
Comment (mosheraboh) [inline fusedrug/data/tokenizer/ops/modular_tokenizer_ops.py:246]: why "_" prefix?

File: fusedrug/data/tokenizer/ops/injector_tokenizer_ops.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
+from fuse.utils import NDict
+
+# from fuse.data import OpBase, get_sample_id
+from fusedrug.data.tokenizer.injectortokenizer.injector_tokenizer import (
+    InjectorTokenizer,
+)
+
+# from fusedrug.data.tokenizer.modulartokenizer.modular_tokenizer import ModularTokenizer
```
Comment (mosheraboh) [inline fusedrug/data/tokenizer/ops/injector_tokenizer_ops.py:8]: remove commented-out imports.

File: fusedrug/data/tokenizer/ops/injector_tokenizer_ops.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+            max_size=max_size,
+            pad_token=pad_token,
+            pad_type_id=pad_type_id,
+            validate_ends_with_eos=validate_ends_with_eos,
+            eos=eos,
+            verbose=verbose,
+            **kwargs,
+        )
+
+        self._input_dim = input_dim
```
Comment (mosheraboh) [inline fusedrug/data/tokenizer/ops/injector_tokenizer_ops.py:70]: are you using it somewhere? the functions below are static.

**Review by mosheraboh — APPROVED** [review]
Looks good.
See few comments inline

File: fusedrug/data/tokenizer/modulartokenizer/modular_tokenizer.py [hunk 5]
```diff
@@ -1025,6 +1031,7 @@ def encode_list(
             on_unknown: (Optional[str], optional): What happens if unknown tokens (i.e. ones mapped to <UNK>) are encountered: 'raise' or 'warn'
             verbose (Optional[int], optional): verbosity level. 0: no notification, 1: warning notification, 2: warning with partial data, 3: warning
                 with full data. Defaults to 1.
+            also_return_split: defaults to False. If set to True, the return value will also contain a list that contains per meta-tokenizer-instruction element of Encoding
```
[the remark this answers is not in the document]
Comment (YoelShoshan) [reply fusedrug/data/tokenizer/modulartokenizer/modular_tokenizer.py:1034]: You don't call this directly, injector_tokenizer_op does it automatically for you.
It's not just for debug
we can't infer it from typed_input_list because we don't know how many tokens will be per tokenizer part (as it's not always 1:1 - there are things like SMILES, and things like cropping/padding)

if we only get the final merged one we can't understand:
1. which tokens we should replace with <SCALARS> , <MASKED_SCALARS>
2. where are the scalars tokens

the only way we can do that externally is by effectively doing the entire logic of modular tokenizer including actual tokenization, padding, cropping, which is both code duplication and will also be slower.
that's why I preferred to allow to return this "internal split" already calculated variable.

If this isn't completely clear yet let's talk

File: fusedrug/data/tokenizer/injectortokenizer/injector_tokenizer.py [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+        ):
+            if tokenizer_type.startswith("SCALARS_"):
+                with_placeholders.append(
+                    "<@TOKENIZER-TYPE=AA>"
+                )  # won't use AA tokens, just an arbitrary one to be able to use a token like <SCALAR>
+
+                if (
+                    tokenizer_type == "SCALARS_LITERALS"
+                ):  # note: masking is only supported in literals (not in "from dict")
+                    values = subseq.split(",")
```
[the remark this answers is not in the document]
Comment (YoelShoshan) [reply fusedrug/data/tokenizer/injectortokenizer/injector_tokenizer.py:74]: yes, the scalars tokenizer require that you split them with ','
if you have an alternative you prefer do suggest.
I will add some description of the expected format in the injector files docstrings

[the remark this answers is not in the document]
Comment (YoelShoshan) [reply fusedrug/data/tokenizer/injectortokenizer/injector_tokenizer.py:74]: added docstrings with format description for both injector_tokenizer.py and injector_tokenizer_op

also renamed InjectorTokenizer to InjectorTokenizerHelpers and stopped inheriting from ModularTokenizer in it because it's misleading, as it's just 2 static helper methods.

[the remark this answers is not in the document]
Comment (YoelShoshan) [reply fusedrug/data/tokenizer/injectortokenizer/injector_tokenizer.py:74]: this is part of the docstrings I've added

```
applies a injector tokenizer

    injector tokenizer builds on top of modular tokenizer.
    its purpose is to build inputs_emb for the model (instead of input_ids)
        this allows to support more advanced inputs beyond token ids, like:
        * scalars inputs
        * embeddings vector within a single input

    supported syntax/format:

    for text following <@TOKENIZER-TYPE=SCALARS_LITERALS> supports the following format:
    ',' separated float values and/or <MASK> tokens - 
        for example: "2.7,3.99,-12.9" or "<MASK><MASK>" or "2.19,<MASK>,3.19,<MASK>"
    
    for text following <@TOKENIZER-TYPE=SCALARS_FROM_DICT> is expected to be a key to the sample NDict
        for example: "blah.boo.banana"  or "data.input.encoder_input"
        note: in SCALARS_FROM_DICT you can't describe masked scalars (outputs) you can only describe inputs

    example usage:

    encoder_input:
    <@TOKENIZER-TYPE=AA><MOLECULAR_WEIGHT_IN_SOME_UNIT><@TOKENIZER-TYPE=SCALARS_LITERALS>0.3<@TOKENIZER-TYPE=AA><BINDING_AFFINITY_NANOMOLAR><@TOKENIZER-TYPE=SCALARS_LITERALS><MASK><@TOKENIZER-TYPE=AA><SEQUENCE_NATURAL_START>ISGGDAIYSSTGRCSLGFNVRSGSTYYFLTAGICTDGATTWWANSARTTVLGTTSGSSFPNNDYGIVRYTNTTIPKDGTVGGQDITSAANATVGMAVTRRGSTTGTISGSVTALNATVNYGGGDVVYGMIRTNVCAEPGDSGGPLYSGTRAIGLTSGGSGNCSSGGTTFFQPVTEALVAYGVSVY<SEQUENCE_NATURAL_END>
    labels:
    <@TOKENIZER-TYPE=AA><MOLECULAR_WEIGHT_IN_SOME_UNIT><@TOKENIZER-TYPE=SCALARS_LITERALS>0.3<@TOKENIZER-TYPE=AA><BINDING_AFFINITY_NANOMOLAR><@TOKENIZER-TYPE=SCALARS_LITERALS>12.4<@TOKENIZER-TYPE=AA><SEQUENCE_NATURAL_START>ISGGDAIYSSTGRCSLGFNVRSGSTYYFLTAGICTDGATTWWANSARTTVLGTTSGSSFPNNDYGIVRYTNTTIPKDGTVGGQDITSAANATVGMAVTRRGSTTGTISGSVTALNATVNYGGGDVVYGMIRTNVCAEPGDSGGPLYSGTRAIGLTSGGSGNCSSGGTTFFQPVTEALVAYGVSVY<SEQUENCE_NATURAL_END>
```

File: fusedrug/data/tokenizer/injectortokenizer/injector_tokenizer.py [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+                #     values = subseq.split(",")
+                #     assert all([x=='<MASK>' for x in values]) #only <MASK> is currently supported
+                #     seq = "<MASKED_SCALAR>" * len(values)
+
+                with_placeholders.append(seq)
+
+            elif tokenizer_type.startswith("VECTORS_"):
+                raise Exception("VECTOR_* are not supported yet")
+            else:
+                with_placeholders.append("<@TOKENIZER-TYPE=" + tokenizer_type + ">")
```
[the remark this answers is not in the document]
Comment (YoelShoshan) [reply fusedrug/data/tokenizer/injectortokenizer/injector_tokenizer.py:103]: I don't think so:

```
sequence = "<@TOKENIZER-TYPE=AA><BLAH><BLAH2>QKPGQAPRLLIYG<@TOKENIZER-TYPE=AA@MAX-LEN=122><BLAH3>SGSDFSDFSFD"
hints_and_subseq = re.split("<@TOKENIZER-TYPE=([^>]*)>", sequence)[1]
```

```
In [6]: hints_and_subseq
Out[6]: ['AA', '<BLAH><BLAH2>QKPGQAPRLLIYG', 'AA@MAX-LEN=122', '<BLAH3>SGSDFSDFSFD']
```

tell me if you still think I miss something here

File: fusedrug/data/tokenizer/injectortokenizer/injector_tokenizer.py [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+
+                    curr_indices = []
+                    curr_data = []
+
+                    for i, val in enumerate(curr_str_data):
+                        if val != "<MASK>":
+                            curr_indices.append(i + prev_index_end + 1)
+                            curr_data.append(float(val))
+                        else:
+                            scalars_masked_indices.append(i + prev_index_end + 1)
```
[the remark this answers is not in the document]
Comment (YoelShoshan) [reply fusedrug/data/tokenizer/injectortokenizer/injector_tokenizer.py:156]: yes, this collects all of the indices (at the level of final tokens) of masked scalars, across the entire sequence.
expected to be empty for labels, and possibly non-empty for encoder_input yeah.

File: fusedrug/data/tokenizer/ops/modular_tokenizer_ops.py (hunk 2 above)
Comment (YoelShoshan) [reply fusedrug/data/tokenizer/ops/modular_tokenizer_ops.py:246]: to hint that it's "private" or "local" - and should not be used as the answer to outside,
just a convention - it's possible that it's a convention I only have with myself :D

File: fusedrug/data/tokenizer/ops/injector_tokenizer_ops.py (hunk 4 above)
Comment (YoelShoshan) [reply fusedrug/data/tokenizer/ops/injector_tokenizer_ops.py:70]: removed it from here and also from multiple related places.
